File: 983_minimum_cost_for_tickets.cpp
class Solution {
public:
    int mincostTickets(vector<int>& x, vector<int>& y) {
        int n = x.size(), i = 0, j = 0;
        vector<int> v(n);
        for (int k = 0; k < n; ++k) {
            while (x[k] - x[i] >= 7) i++;
            while (x[k] - x[j] >= 30) j++;
            v[k] = min({(k ? v[k - 1] : 0) + y[0], (i ? v[i - 1] : 0) + y[1], (j ? v[j - 1] : 0) + y[2]});
        }
        return v[n - 1];
    }
};

File: 983_minimum_cost_for_tickets.java
class Solution {
    public int mincostTickets(int[] t, int[] f) {
        int k = t.length, u = 0, v = 0;
        int[] w = new int[k];
        for (int e = 0; e < k; e++) {
            while (t[e] - t[u] >= 7) u++;
            while (t[e] - t[v] >= 30) v++;
            w[e] = Math.min((e > 0 ? w[e - 1] : 0) + f[0], 
                            Math.min((u > 0 ? w[u - 1] : 0) + f[1], 
                                     (v > 0 ? w[v - 1] : 0) + f[2]));
        }
        return w[k - 1];
    }
}
